Performance optimizations for JPEG decompression

xqms opened the pull request:
This PR reduces JPEG decompression time by roughly factor 2 for larger images.

First of all, we use the TurboJPEG library directly rather than through OpenCV's imdecode() method. TurboJPEG is part of libjpeg-turbo and thus installed on most systems anyway. Using TurboJPEG has two advantages:

 * RGB/BGR conversions can be fused with the YUV to RGB conversion that is required during JPEG decompression anyway.
 * TurboJPEG's API makes it easy to achieve zero-copy operation.

We remove two data copies: One during RGB/BGR conversion and one during the final conversion to `sensor_msgs::Image`.

On my system, this PR allows me to decode a 4K JPEG stream (encoded externally from ROS) at 30 Hz without a problem. Without the PR, I could at most achieve 20 Hz.

I tested mostly the stream configuration I had available. I'll do further RGB/BGR tests soon. Since there are no unit tests for `compressed_image_transport` (or I didn't find them), this requires careful attention.

xqms committed:
find & link to TurboJPEG library

xqms committed:
compressed_image_transport: use TurboJPEG for JPEG images & avoid copies

Before, we copied the pixel data once during the color conversions
and once during sending (CvImage::toImageMsg()). Using the TurboJPEG
API, we can directly write into the resulting message.

On my system, this makes it possible to process 4K @ 30 Hz using a single
thread.

xqms committed:
compressed_image_transport: fix compilation issues on Kinetic

xqms committed:
compressed_image_transport: fix TurboJPEG linking on Xenial

peci1 commented:
This PR breaks ABI by adding the `tjhandle tj_` member variable. Would there be an ABI-compatible way to implement this?

Otherwise, I agree a faster decode engine would be useful - currently, displaying a single 1600x7000 image stream in rviz on Steam Deck results in rviz running with 2 FPS, which is practically useless.

xqms commented:
Hmm, good point. But is ABI a practical concern here? I would think most people use `compressed_image_transport` as a plugin and in that case the plugin base class is the externally visible ABI.

Keeping the ABI of the `CompressedImageSubscriber` class the same is very difficult, since the destructor is defined in the header - we have no possibility of doing any cleanup (e.g. freeing the `tj_` handle) on destruction. With a lot of hacking we could maybe change the generated code for the reconfigure class to stash away the handle there... But that seems very fragile.

So I would propose a solution in the middle: We create a new class (FastCompressedImageSubscriber or similar), which uses turbojpeg. The plugin XML files point to this new class. Anyone who is using the library `compressed_image_subscriber` directly (again, who would do that?) would still use the old class.

I don't have a lot of time to work on this, though. For my use cases I have a much-more performant alternative now (https://github.com/xqms/imgui_ros - documentation is coming!), so the pressing need for this on my side is now gone ;) I'm happy to make one more revision of this PR, though - or help anyone else who would be willing to take this up.

peci1 commented:
> But is ABI a practical concern here? I would think most people use `compressed_image_transport` as a plugin and in that case the plugin base class is the externally visible ABI.

Good point. I've actually tried to use the compressed transport as a direct library, but I haven't succeded - the API is just useless for any other use-case than `ImageSubscriber`. The decode/encode functions themselves are not factored out from the publishing code, so the idea to just use the decoder's C++ API did not work.

So maybe it is actually ok here to break ABI. Sorry for the confusion.

ijnek commented:
I recently took over maintenance for the ROS 2 branches, but the maintainer for ROS 1 branches have been out for many years. I'm happy to merge this in if @peci1 or someone else could review.

peci1 commented on compressed_image_transport/CMakeLists.txt:
```diff
@@ -1,7 +1,10 @@
 cmake_minimum_required(VERSION 3.0.2)
 project(compressed_image_transport)
 
+set(CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/modules" ${CMAKE_MODULE_PATH})
+
 find_package(OpenCV REQUIRED)
+find_package(TurboJPEG REQUIRED)
 find_package(catkin REQUIRED COMPONENTS cv_bridge dynamic_reconfigure image_transport)
```
```suggestion
find_package(OpenCV REQUIRED)
find_package(catkin REQUIRED COMPONENTS cv_bridge dynamic_reconfigure image_transport)

find_package(PkgConfig REQUIRED)
pkg_check_modules(TurboJPEG REQUIRED libturbojpeg)
```

peci1 replied:
Ubuntu, Fedora, Debian, Arch and probably all other distros provide a pkgconfig file, so let's use it.

xqms replied:
Ubuntu 16.04 does not provide a pkgconfig file. But I guess since Kinetic is EOL we don't care about 16.04.

peci1 commented on compressed_image_transport/modules/FindTurboJPEG.cmake:
```diff
@@ -0,0 +1,21 @@
+# Find TurboJPEG
+#
+# The following are set after configuration is done:
+#  TurboJPEG_FOUND
+#  TurboJPEG_INCLUDE_DIRS
+#  TurboJPEG_LIBRARIES
+
+include(FindPackageHandleStandardArgs)
+
+# On Ubuntu Xenial, a missing symlink means we have to specify the name exactly
+find_library(TurboJPEG_LIBRARY libturbojpeg.so.0 turbojpeg)
+find_path(TurboJPEG_INCLUDE_DIR turbojpeg.h)
+
+find_package_handle_standard_args(TurboJPEG
+  REQUIRED_VARS TurboJPEG_LIBRARY TurboJPEG_INCLUDE_DIR
+)
+
+if(TurboJPEG_FOUND)
+  set(TurboJPEG_INCLUDE_DIRS "${TurboJPEG_INCLUDE_DIR}")
+  set(TurboJPEG_LIBRARIES "${TurboJPEG_LIBRARY}")
+endif()
```
```suggestion
```
Not needed (substituted by pkg_config).

peci1 commented on compressed_image_transport/include/compressed_image_transport/compressed_subscriber.h:
```diff
@@ -61,11 +64,14 @@ class CompressedSubscriber : public image_transport::SimpleSubscriberPlugin<sens
   boost::shared_ptr<ReconfigureServer> reconfigure_server_;
   Config config_;
   int imdecode_flag_;
+  tjhandle tj_;
```
I tested performance without the class member (initing decompressor in each callback) and the performance loss wasn't that big. So if we'd want to keep ABI, this could be also a possibility.

xqms replied:
I guess that is up to you. I still think nobody is using `compressed_image_transport` directly, so there should be no impact of an ABI change. But I'm also fine with moving it to the callback.

peci1 replied:
Feel free to leave it as it is. I think ABI compatibility here is not a key aspect as we're talking about a plugin.

peci1 commented on compressed_image_transport/package.xml:
```diff
@@ -17,10 +17,12 @@
   <run_depend>cv_bridge</run_depend>
   <run_depend>dynamic_reconfigure</run_depend>
   <run_depend>image_transport</run_depend>
+  <run_depend>libturbojpeg</run_depend>
```
I suggest switchin to package format 2 and declaring this only as `<exec_depend>` - otherwise all dependent packages would need to link to turbojpeg, which is not needed.

xqms replied:
Good point, will do so.

xqms replied:
Actually, at the moment we `#include <turbojpeg.h>` in `compressed_subscriber.h`, which makes it a build export dependency, right?

We could maybe replace the `tjhandle` with a `void*` and avoid the include in the header. Or have a pimpl-style `std::unique_ptr<Private>` member, that way we can define the `tjhandle` in the cpp file.

peci1 replied:
I think the reasoning here is the same as with the ABI compatibility - technically, it would add a build-export dependency if someone linked to the plugin, but doing so is highly improbable...

xqms replied:
Ok!

peci1 commented on compressed_image_transport/src/compressed_subscriber.cpp:
```diff
@@ -82,11 +93,118 @@ void CompressedSubscriber::shutdown()
                                             const Callback& user_cb)
 
 {
+  // Parse format field
+  std::string source_encoding;
+  std::string compression;
+  {
+    const size_t split_pos = message->format.find(';');
+    if (split_pos != std::string::npos)
+    {
+      source_encoding = message->format.substr(0, split_pos);
+      compression = message->format.substr(split_pos);
+    }
+  }
 
+  // Try TurboJPEG first (if the first bytes look like JPEG)
+  if(message->data.size() > 4 && message->data[0] == 0xFF && message->data[1] == 0xD8)
+  {
+    sensor_msgs::ImagePtr decoded = decompressJPEG(message->data, source_encoding, message->header);
+    if(decoded)
+    {
+      user_cb(decoded);
+      return;
+    }
+  }
+
+  // Otherwise, try our luck with OpenCV.
```
```suggestion
  // Parse format field
  std::string image_encoding;
  std::string compressed_encoding;
  {
    const size_t split_pos = message->format.find(';');
    if (split_pos != std::string::npos)
    {
      image_encoding = message->format.substr(0, split_pos);
      compressed_encoding = message->format.substr(split_pos);
    }
  }

  // Try TurboJPEG first (if the first bytes look like JPEG)
  if(message->data.size() > 4 && message->data[0] == 0xFF && message->data[1] == 0xD8)
  {
    sensor_msgs::ImagePtr decoded = decompressJPEG(message->data, image_encoding, message->header);
    if(decoded)
    {
      user_cb(decoded);
      return;
    }
  }

  // Otherwise, try our luck with OpenCV.
```

Do not unnecessarily rename existing variables.

peci1 commented on compressed_image_transport/src/compressed_subscriber.cpp:
```diff
@@ -98,8 +216,7 @@ void CompressedSubscriber::internalCallback(const sensor_msgs::CompressedImageCo
     // Assign image encoding string
-    const size_t split_pos = message->format.find(';');
-    if (split_pos==std::string::npos)
+    if (source_encoding.empty())
```
```suggestion
    if (image_encoding.empty())
```

peci1 commented on compressed_image_transport/src/compressed_subscriber.cpp:
```diff
@@ -116,37 +233,34 @@ void CompressedSubscriber::internalCallback(const sensor_msgs::CompressedImageCo
-      std::string image_encoding = message->format.substr(0, split_pos);
-
-      cv_ptr->encoding = image_encoding;
+      cv_ptr->encoding = source_encoding;
```
```suggestion
      cv_ptr->encoding = image_encoding;
```

peci1 commented on compressed_image_transport/src/compressed_subscriber.cpp:
```diff
@@ -116,37 +233,34 @@ void CompressedSubscriber::internalCallback(const sensor_msgs::CompressedImageCo
+      cv_ptr->encoding = source_encoding;
 
-      if ( enc::isColor(image_encoding))
+      if ( enc::isColor(source_encoding))
```
```suggestion
      if ( enc::isColor(image_encoding))
```

peci1 commented on compressed_image_transport/src/compressed_subscriber.cpp:
```diff
@@ -116,37 +233,34 @@ void CompressedSubscriber::internalCallback(const sensor_msgs::CompressedImageCo
       {
-        std::string compressed_encoding = message->format.substr(split_pos);
-        bool compressed_bgr_image = (compressed_encoding.find("compressed bgr") != std::string::npos);
+        bool compressed_bgr_image = (compression.find("compressed bgr") != std::string::npos);
```
```suggestion
        bool compressed_bgr_image = (compressed_encoding.find("compressed bgr") != std::string::npos);
```

peci1 commented on compressed_image_transport/src/compressed_subscriber.cpp:
```diff
@@ -116,37 +233,34 @@ void CompressedSubscriber::internalCallback(const sensor_msgs::CompressedImageCo
         {
           // if necessary convert colors from bgr to rgb
-          if ((image_encoding == enc::RGB8) || (image_encoding == enc::RGB16))
+          if ((source_encoding == enc::RGB8) || (source_encoding == enc::RGB16))
```
```suggestion
          if ((image_encoding == enc::RGB8) || (image_encoding == enc::RGB16))
```

peci1 commented on compressed_image_transport/src/compressed_subscriber.cpp:
```diff
@@ -116,37 +233,34 @@ void CompressedSubscriber::internalCallback(const sensor_msgs::CompressedImageCo
             cv::cvtColor(cv_ptr->image, cv_ptr->image, CV_BGR2RGB);
 
-          if ((image_encoding == enc::RGBA8) || (image_encoding == enc::RGBA16))
+          if ((source_encoding == enc::RGBA8) || (source_encoding == enc::RGBA16))
```
```suggestion
          if ((image_encoding == enc::RGBA8) || (image_encoding == enc::RGBA16))
```

peci1 commented on compressed_image_transport/src/compressed_subscriber.cpp:
```diff
@@ -116,37 +233,34 @@ void CompressedSubscriber::internalCallback(const sensor_msgs::CompressedImageCo
             cv::cvtColor(cv_ptr->image, cv_ptr->image, CV_BGR2RGBA);
 
-          if ((image_encoding == enc::BGRA8) || (image_encoding == enc::BGRA16))
+          if ((source_encoding == enc::BGRA8) || (source_encoding == enc::BGRA16))
```
```suggestion
          if ((image_encoding == enc::BGRA8) || (image_encoding == enc::BGRA16))
```

peci1 commented on compressed_image_transport/src/compressed_subscriber.cpp:
```diff
@@ -116,37 +233,34 @@ void CompressedSubscriber::internalCallback(const sensor_msgs::CompressedImageCo
         {
           // if necessary convert colors from rgb to bgr
-          if ((image_encoding == enc::BGR8) || (image_encoding == enc::BGR16))
+          if ((source_encoding == enc::BGR8) || (source_encoding == enc::BGR16))
```
```suggestion
          if ((image_encoding == enc::BGR8) || (image_encoding == enc::BGR16))
```

peci1 commented on compressed_image_transport/src/compressed_subscriber.cpp:
```diff
@@ -116,37 +233,34 @@ void CompressedSubscriber::internalCallback(const sensor_msgs::CompressedImageCo
             cv::cvtColor(cv_ptr->image, cv_ptr->image, CV_RGB2BGR);
 
-          if ((image_encoding == enc::BGRA8) || (image_encoding == enc::BGRA16))
+          if ((source_encoding == enc::BGRA8) || (source_encoding == enc::BGRA16))
```
```suggestion
          if ((image_encoding == enc::BGRA8) || (image_encoding == enc::BGRA16))
```

peci1 commented on compressed_image_transport/src/compressed_subscriber.cpp:
```diff
@@ -116,37 +233,34 @@ void CompressedSubscriber::internalCallback(const sensor_msgs::CompressedImageCo
             cv::cvtColor(cv_ptr->image, cv_ptr->image, CV_RGB2BGRA);
 
-          if ((image_encoding == enc::RGBA8) || (image_encoding == enc::RGBA16))
+          if ((source_encoding == enc::RGBA8) || (source_encoding == enc::RGBA16))
```
```suggestion
          if ((image_encoding == enc::RGBA8) || (image_encoding == enc::RGBA16))
```

peci1 reviewed (changes requested):
I went through this PR and locally tested it.

It works well on BGR, MONO8 and DEPTH (MONO16 PNG) images - the decoded output is the same as in the original version, and performance is better (except for the PNG images, where it is the same).

In my case, I tested performance with images 1616x7392 @ 6 Hz (a tall 6-camera merge from a synchronized omnicam). Playing with RTF set to 1.0, CPU usage of republish node decreases from ~60% to ~50%. When playing with RTF 10.0, the republish node manages republishing ~15 images/s, while the original implementation manages ~8 images/s.

I have some code comments that I'd like to be addressed. @xqms are you willing to still work on this PR, or should I take over it?

xqms committed:
Switch to pkgconfig

Co-authored-by: Martin Pecka <peci1@seznam.cz>

xqms committed:
Do not rename variables

Co-authored-by: Martin Pecka <peci1@seznam.cz>

peci1 commented:
In the meantime, I've discovered a transport plugin doing the encoding part via turbojpeg: https://github.com/wep21/turbojpeg_compressed_image_transport/blob/main/src/turbojpeg_compressed_publisher.cpp . Would you be interested in doing that part, too? Maybe in a followup PR?

xqms committed:
compressed_image_transport: package.xml: format 2 + more specific deps

cv_bridge, dynamic_reconfigure, and libturbojpeg are only used internally
and we do not expect packages depending on compressed_image_transport will
need them, since they use the plugin interface to load & execute the code.

xqms commented:
> In the meantime, I've discovered a transport plugin doing the encoding part via turbojpeg: https://github.com/wep21/turbojpeg_compressed_image_transport/blob/main/src/turbojpeg_compressed_publisher.cpp . Would you be interested in doing that part, too? Maybe in a followup PR?

Hmm, I currently don't have high-bandwidth JPEG publisher using `image_transport` in use, so this would be quite low on my priority list. I may get around to doing it sometime next year.

peci1 reviewed (approved):
I've retested this PR locally and everything works.

All my comments have been addressed, except for a dangling empty file `modules/FindTurboJPEG.cmake`, which is a remnant of the previous version and should be completely deleted.

ijnek commented:
With the approval, I'm going to go ahead and merge this. Thanks for the contribution @xqms, sorry for the very long wait and thanks for reviewing @peci1!

peci1 commented:
Awesome! Thanks @xqms for working on this!

Could I also ask for a Melodic backport?

ijnek commented:
Just realized the melodic version of image_transport_plugins points to the indigo-devel branch :open_mouth: 

I'm going to guess that the addition of libturbojpeg may cause issues in some users if they don't have libturbojpeg available on their system. So I'm hesitant for this backport. What do you think @peci1 ?

peci1 commented:
> points to the indigo-devel branch

That's quite common. There weren't too many breaking changes between indigo and melodic, so many packages share the same codebase for both. However, indigo is EOL for quite long, so I don't see a problem there. Regarding melodic, the situation is the same as for noetic - if somebody does not have libjpeg-turbo installed, he will need to add this dependency. However, that is also a common case. APT will take care of that in case the user runs on the binary packages. In case of from-source builds, the users should anyways do `rosdep install` after each source update, so this case should also be covered. And, last, libjpeg-turbo is available for all platforms of interest. I've personally checked this PR on melodic and it works fine.

ijnek commented:
I'm comfortable with how things are done in ROS 2, but not too much ROS 1. Assuming things are done the same way, you'd never want to backport changes that add extra dependencies, even if you can confirm it works on the listed supported platforms. That way it caters for those on platforms that aren't listed but have successfully used the package up until the change.

That said, as you said libturbojpeg does seem covered pretty well across different distributions:
https://github.com/ros/rosdistro/blob/9cfaef7339b94ee28daa8ed1af69be870a37ab0d/rosdep/base.yaml#L5837-L5861

---

In regards to the indigo-devel branch, would you agree with branching off melodic-devel from indigo-devel and applying the changes there so we can ensure indigo-devel stays in a working state.

Looking forward to hearing your thoughts.

peci1 commented:
> you'd never want to backport changes that add extra dependencies

But this happens regularly when I update my system. Moreover, I don't understand the difference between noetic and melodic in this regard. Why would introducing a new dependency be a problem in Melodic and not a problem in Noetic?

> branching off melodic-devel

This might make sense as some distributions for kinetic do not have the PkgConfig for libjpeg-turbo. However, I'm not sure about the policy when a devel branch should be branched off, so that's probably more a maintainer-level decision.

peci1 commented:
A recent example of a dependency being added to Moveit "in the middle" of a released ROS version:

https://github.com/ros-planning/moveit/pull/2547/files#diff-a572b9bd78f220384c2c0ede5c92744f22f3bcc7bb6962847725f6507ef22cd4

and being backported to melodic here: https://github.com/ros-planning/moveit/pull/2651 .

ijnek commented:
> Why would introducing a new dependency be a problem in Melodic and not a problem in Noetic?

This argument applies to merging in the changes to the noetic branch too. I just only thought of it after being requested to backport.

Given the links to the examples in moveit, I'm leaning towards branching-off melodic-devel and backporting the changes there as requested. Does that sound fine?

peci1 commented:
> branching-off melodic-devel and backporting the changes there as requested. Does that sound fine?

Perfect! Thanks.

ijnek commented:
@Mergifyio backport melodic-devel

peci1 commented:
Looks promising.

---

Ahh, now I see there is still the empty cmake file in this PR, could you please remove it, too?

ijnek commented:
Updating rosdistro for melodic here: https://github.com/ros/rosdistro/pull/35733

ijnek commented:
> Ahh, now I see there is still the empty cmake file in this PR, could you please remove it, too?

https://github.com/ros-perception/image_transport_plugins/pull/124
